text/0036-async-testbench-functions.md

```diff
@@ -0,0 +1,276 @@
+> A way to attach clock domain information to interfaces is desireable, but out of scope for this RFC.
+
+Using this stream interface, let's consider a colorspace converter accepting a stream of RGB values and outputting a stream of YUV values:
+
+```python
+class RGBToYUVConverter(Component):
+    input: In(StreamSignature(RGB888))
+    output: Out(StreamSignature(YUV888))
+```
+
+A testbench could then look like this:
+
+```python
+async def test_rgb(sim, r, g, b):
+    rgb = {'r': r, 'g': g, 'b': b}
+    await stream_send(sim, dut.input, rgb)
+    yuv = await stream_recv(sim, dut.output)
+
+    print(rgb, yuv)
+
+async def testbench(sim):
+    await test_rgb(sim, 0, 0, 0)
+    await test_rgb(sim, 255, 0, 0)
+    await test_rgb(sim, 0, 255, 0)
+    await test_rgb(sim, 0, 0, 255)
+    await test_rgb(sim, 255, 255, 255)
+```
+
+Since `stream_send()` and `stream_recv()` invokes `sim.get()` and `sim.set()` that in turn will invoke the appropriate value conversions for a value castable (here `data.View`), it is general enough to work for streams with arbitrary shapes.
+
+`Tick()` and `Delay()` are replaced by `sim.tick()` and `sim.delay()` respectively.
+In addition, `sim.changed()` and `sim.edge()` is introduced that allows creating triggers from arbitrary signals.
+
+`sim.tick()` return a domain trigger object that can be made conditional through `.until()` or repeated through `.repeat()`.
+
+`sim.delay()`, `sim.changed()` and `sim.edge()` return a combinable trigger object that can be used to add additional triggers.
+
+`Active()` and `Passive()` are replaced by an `background=False` keyword argument to `.add_testbench()`.
+Processes created through `.add_process()` are always created as background processes.
+To allow a background process to ensure an operation is finished before end of simulation, `sim.critical()` is introduced, which is used as a context manager:
+
+```python
+async def packet_reader(sim, stream):
+    while True:
+        # Wait until stream has valid data.
+        await sim.tick().until(stream.valid)
+
+        # Ensure simulation doesn't end in the middle of a packet.
+        async with sim.critical():
+            packet = await stream.read_packet()
+            print('Received packet:', packet.hex(' '))
+```
+
+When a combinable trigger object is awaited, it'll return the value(s) of the trigger(s), and it can also be used as an async generator to repeatedly await the same trigger.
+Multiple triggers can be combined.
+Consider the following examples:
+
+Combinational adder as a process:
+```python
+a = Signal(); b = Signal(); o = Signal()
+async def adder(sim):
+    async for a_val, b_val in sim.changed(a, b):
+        await sim.set(o, a_val + b_val)
+sim.add_process(adder)
+```
+
+DDR IO buffer as a process:
+```python
+o = Signal(2); pin = Signal()
+async def ddr_buffer(sim):
+    while True: # could be extended to pre-capture next `o` on posedge
+        await sim.negedge()
```

Comment:
I think this isn't valid code--needs to be `await sim.negedge(clk)` or something.